packages/vc-export/src/KiltAttestationProofV1.ts

```diff
@@ -0,0 +1,633 @@
+ *
+ * @param credentialInput Verifiable Credential to verify proof against.
+ * @param proof KiltAttestationProofV1 proof object to be verified. Any proofs embedded in the credentialInput are stripped and ignored.
+ * @param opts Additional parameters.
+ * @param opts.api A polkadot-js/api instance connected to the blockchain network on which the credential is anchored.
+ * @returns Object indicating whether proof could be verified.
+ */
+export async function verifyProof(
+  credentialInput: VerifiableCredential,
+  proof: KiltAttestationProofV1,
+  opts: { api?: ApiPromise } = {}
+): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { proof: _, ...credential } = credentialInput
+  // 0. check proof structure
+  validateStructure(proof)
+  // 1 - 3. check credential structure
+  validateCredentialStructure(credential)
+  const { nonTransferable, credentialStatus, credentialSubject, issuer } =
+    credential
+  validateUri(issuer, 'Did')
+  // 4. check nonTransferable
+  if (nonTransferable !== true)
+    throw new CredentialMalformedError('nonTransferable must be true')
+  // 5. check credentialStatus
+  if (credentialStatus.type !== KILT_REVOCATION_STATUS_V1_TYPE)
+    throw new CredentialMalformedError(
+      `credentialStatus must have type ${KILT_REVOCATION_STATUS_V1_TYPE}`
+    )
+  const { assetInstance, assetNamespace, assetReference } = Caip19.parse(
+    credentialStatus.id
+  )
+  const expectedAttestationId = credential.id.substring(
+    KILT_CREDENTIAL_IRI_PREFIX.length
+  )
+  if (
+    assetNamespace !== 'kilt' ||
+    assetReference !== 'attestation' ||
+    assetInstance !== expectedAttestationId
+  ) {
+    throw new Error(
+      `credentialStatus.id must end on 'kilt:attestation/${expectedAttestationId} in order to be verifiable with this proof`
+    )
+  }
+  // 6. json-ld expand credentialSubject
+  const expandedContents = jsonLdExpandCredentialSubject(credentialSubject)
+  // 7. Transform to normalized statments and hash
+  const { statements, digests } = normalizeClaims(expandedContents)
+  if (statements.length !== proof.salt.length)
+    throw new Error(
+      'Violated expectation: number of normalized statements === number of salts'
+    )
+  // 8-9. Re-compute commitments
+  const commitments = makeCommitments(
+    digests,
+    proof.salt.map((salt) => base58Decode(salt))
+  )
+  // 10. Assert commitments are in proof
+  commitments.forEach((recomputed, index) => {
+    if (!proof.commitments.includes(base58Encode(recomputed)))
+      throw new Error(
+        `No commitment for statement with digest ${u8aToHex(
+          digests[index]
+        )} and salt ${proof.salt[index]}`
+      )
+  })
+  // 11. Compute root hash
+  const rootHash = calculateRootHash(credential, proof)
+  // 12. Compare against credential id
+  if (credentialIdFromRootHash(rootHash) !== credential.id)
+    throw new Error('root hash not verifiable')
```

Comment:
I guess we want to throw credential or proof errors, right?

Reply:
673a7a95fe712b267c66fbd3d9610f246f66c6b5